go/test/endtoend/vtgate/tabletbalancer/balancer_test.go

```diff
@@ -0,0 +1,483 @@
+/*
+Copyright 2025 The Vitess Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package tabletbalancer
+
+import (
+	"context"
+	"fmt"
+	"testing"
+	"time"
+
+	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/require"
+
+	"vitess.io/vitess/go/mysql"
+	"vitess.io/vitess/go/test/endtoend/cluster"
+	"vitess.io/vitess/go/vt/vtgate/planbuilder/plancontext"
+)
+
+// Helper functions for e2e testing of balancer modes
+
+// executeReplicaQueries executes queries against replicas and verifies they succeed
+func executeReplicaQueries(t *testing.T, conn *mysql.Conn, numQueries int) map[int64]int {
+	t.Helper()
+
+	// Use @replica to target replica tablets
+	_, err := conn.ExecuteFetch("USE @replica", 1, false)
+	require.NoError(t, err)
+
+	counts := make(map[int64]int)
+
+	// Execute queries and verify they succeed
+	for range numQueries {
+		res, err := conn.ExecuteFetch("SELECT @@server_id, id FROM balancer_test LIMIT 1", 10, false)
+		require.NoError(t, err)
+		require.Equal(t, 1, len(res.Rows), "expected one row from balancer_test")
+
+		serverID, err := res.Rows[0][0].ToInt64()
+		require.NoError(t, err)
+		counts[serverID]++
+	}
+
+	return counts
+}
+
+func getTabletIDFromServerID(t *testing.T, tablets []*cluster.Vttablet) map[string]int64 {
```

Comment:
Nit, but mapTabletAliasToMySQLServerID would be a better name. It might also be worth using @@server_uuid instead of id.